ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -361,6 +361,99 @@ public class Manager: NSObject, HTTPDownloadDelegate, UIGestureRecognizerDelegat
     }
     return nil
   }
+    
+  // MARK: - Adhoc keyboards
+
+  
+  public func parseKMP(_ folder: URL, complete: @escaping (Int) -> Void) {
+    do {
+      var path = folder
+      path.appendPathComponent("kmp.json")
+      let data = try Data(contentsOf: path, options: .mappedIfSafe)
+      let jsonResult = try JSONSerialization.jsonObject(with: data, options: .mutableLeaves)
+      if let jsonResult = jsonResult as? [String:AnyObject] {
+        if let keyboards = jsonResult["keyboards"] as? [[String:AnyObject]] {
+          for k in keyboards {
+            let name = k["name"] as! String
+            let keyboardID = k["id"] as! String
+            let version = k["version"] as! String
+            let osk = k["oskFont"] as! String
+            let font = k["displayFont"] as! String
+            let displayFont = Font(filename: font)
+            let oskFont = Font(filename: osk)
+
+            //TODO: handle errors if languages do not exist
+            var languageName = ""
+            var languageId = ""
+            
+            var installableKeyboards : [InstallableKeyboard] = []
+            if let langs = k["languages"] as? [[String:String]] {
+              for l in langs {
+                languageName = l["name"]!
+                languageId = l["id"]!
+                
+                installableKeyboards.append( InstallableKeyboard(id: keyboardID, name: name, languageID: languageId, languageName: languageName, version: version, isRTL: false, font: displayFont, oskFont: oskFont, isCustom: false))
```

Comment:
at the moment on Android, I'm assuming all KMP distributed keyboards  are ad-hoc where `isCustom = true`

Reply:
Heads up. @darcywong00 That will need to change in future as we do plan to move to 100% KMP distribution. Worth planning ahead.